### ISSUE TITLE
Developers can use System.Text.Json to serialize type hierarchies securely

### ISSUE THREAD

**Issue by eiriktsarpalis** [dotnet/runtime#63747]
## Background and Motivation

Serializing type hierarchies (aka polymorphic serialization) has been a feature long requested by the community (cf. #29937, #30083). We took a stab at [producing a polymorphism feature](https://github.com/dotnet/runtime/pull/53882) during the .NET 6 development cycle, ultimately though the feature was cut since we eventually reached the conclusion that unconstrained polymorphism does not meet our security bar, under any circumstance.

For context, polymorphic serialization has long been associated with security vulnerabilities. More specifically, unconstrained polymorphic serialization can result in accidental data disclosure and unconstrained polymorphic deserialization can result in remote code execution when used with untrusted input. See the [BinaryFormatter security guide](https://docs.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide) for an explanation of such vulnerabilities.

We acknowledge that serializing type hierarchies is an important feature, and we are committed to delivering a _secure_ implementation in a future release of System.Text.Json. This implies that we will be releasing a brand of polymorphism that is restricted by design, requiring users to explicitly opt-in to supported subtypes.

## Basic Polymorphism

At the core of the design is the introduction of the `JsonDerivedType` attribute:
```csharp
[JsonDerivedType(typeof(Derived))]
public class Base
{
    public int X { get; set; }
}

public class Derived : Base
{
    public int Y { get; set; }
}
```
This configuration enables polymorphic serialization for `Base`, specifically when the runtime type is `Derived`:
```csharp
Base value = new Derived();
JsonSerializer.Serialize<Base>(value); // { "X" : 0, "Y" : 0 }
```
Note that this does not enable polymorphic _deserialization_ since the payload would roundtripped as `Base`:
```C#
Base value = JsonSerializer.Deserialize<Base>(@"{ ""X"" : 0, ""Y"" : 0 }");
value is Derived; // false
```

## Polymorphism using Type Discriminators

To enable polymorphic _deserialization_, users need to specify a _type discriminator_ for the derived class:
```C#
[JsonDerivedType(typeof(Base), typeDiscriminatorId: "base")]
[JsonDerivedType(typeof(Derived), typeDiscriminatorId: "derived")]
public class Base
{
    public int X { get; set; }
}

public class Derived : Base
{
    public int Y { get; set; }
}
```
Which will now emit JSON along with type discriminator metadata:
```C#
Base value = new Derived();
JsonSerializer.Serialize<Base>(value); // { "$type" : "derived", "X" : 0, "Y" : 0 }
```
which can be used to deserialize the value polymorphically:
```C#
Base value = JsonSerializer.Deserialize<Base>(@"{  ""$type"" : ""derived"", ""X"" : 0, ""Y"" : 0 }");
value is Derived; // true
```
Type discriminator identifiers can also be integers, so the following form is valid:
```C#
[JsonDerivedType(typeof(Derived1), 0)]
[JsonDerivedType(typeof(Derived2), 1)]
[JsonDerivedType(typeof(Derived3), 2)]
public class Base { }

JsonSerializer.Serialize<Base>(new Derived2()); // { "$type" : 1, ... }
```

## Mixing and matching configuration

It is possible to mix and match type discriminator configuration for a given type hierarchy
```C#
[JsonPolymorphic]
[JsonDerivedType(typeof(Derived1))]
[JsonDerivedType(typeof(Derived2), "derived2")]
[JsonDerivedType(typeof(Derived3), 3)]
public class Base
{
    public int X { get; set; }
}
```
resulting in the following serializations:
```C#
var json1 = JsonSerializer.Serialize<Base>(new Derived1()); // { "X" : 0, "Y" : 0 }
var json2 = JsonSerializer.Serialize<Base>(new Derived2()); // { "$type" : "derived2", "X" : 0, "Z" : 0 }
```

## Customizing Type Disciminators

It is possible to customize the property name of the type discriminator metadata like so:

```C#
[JsonPolymorphic(CustomTypeDiscriminatorPropertyName = "$case")]
[JsonDerivedType(typeof(Derived1), "derived1")]
public class Base
{
    public int X { get; set; }
}
```

resulting in the following JSON:

```csharp
JsonSerializer.Serialize<Base>(new Derived1()); // { "$case" : "derived1", "X" : 0, "Y" : 0 }
```

## Unknown Derived Type Handling

Consider the following type hierarchy:
```C#
[JsonDerivedType(typeof(DerivedType1))]
public class Base { }
public class DerivedType1 : Base { }
public class DerivedType2 : Base { }
```
Since the configuration does not explicitly opt in support for `DerivedType2`, attempting to serialize instances of `DerivedType2` as `Base` will result in a runtime exception:
```C#
JsonSerializer.Serialize<Base>(new DerivedType2()); // throws NotSupportedException
```
The default behavior can be tweaked using the `JsonUnknownDerivedTypeHandling` enum, which can be specified like so:
```C#
[JsonPolymorphic(UnknownDerivedTypeHandling = JsonUnknownDerivedTypeHandling.FallBackToBaseType)]
[JsonDerivedType(typeof(DerivedType1))]
public class Base { }

JsonSerializer.Serialize<Base>(new DerivedType2()); // serialize using the contract for `Base`
```
The `FallBackToNearestAncestor` setting can be used to fall back to the contract of nearest declared derived type:
```C#
[JsonPolymorphic(UnknownDerivedTypeHandling = JsonUnknownDerivedTypeHandling.FallBackToNearestAncestor)]
[JsonDerivedType(typeof(MyDerivedClass)]
public interface IMyInterface { }
public class MyDerivedClass : IMyInterface { }

public class TestClass : MyDerivedClass { }

JsonSerializer.Serialize<IMyInterface>(new TestClass()); // serializes using the contract for `MyDerivedClass`
```
It should be noted that falling back to the nearest ancestor admits the possibility of diamond ambiguity:
```C#
[JsonPolymorphic(UnknownDerivedTypeHandling = JsonUnknownDerivedTypeHandling.FallBackToNearestAncestor)]
[JsonDerivedType(typeof(MyDerivedClass)]
public interface IMyInterface { }

public class IMyDerivedInterface : IMyInterface { }
public class MyDerivedClass : IMyInterface { }

public class TestClass : MyDerivedClass, IMyDerivedInterface { }

JsonSerializer.Serialize<IMyInterface>(new TestClass()); // throws NotSupportedException
```

## Configuring Polymorphism via the Contract model

For use cases where attribute annotations are impractical or impossible (large domain models, cross-assembly hierarchies, hierarchies in third-party dependencies, etc.), it should still be possible to configure polymorphism using the [JSON contract model](https://github.com/dotnet/runtime/issues/63686):
```C#
public class MyPolymorphicTypeResolver : DefaultJsonTypeInfoResolver
{
    public JsonTypeInfo GetTypeInfo(Type type, JsonSerializerOptions options)
    {
        JsonTypeInfo jsonTypeInfo = base.GetTypeInfo(type, options);
        if (jsonTypeInfo.Type == typeof(Base))
        {
            jsonTypeInfo.PolymorphismOptions =
                new JsonPolymorphismOptions
                {
                     TypeDiscriminatorPropertyName = "_case",
                     UnknownDerivedTypeHandling = JsonUnknownDerivedTypeHandling.FallBackToNearestAncestor,
                     DerivedTypes =
                     {
                          new JsonDerivedType(typeof(DerivedType1)),
                          new JsonDerivedType(typeof(DerivedType2), "derivedType2"),
                          new JsonDerivedType(typeof(DerivedType3), 42),
                     }
               }
        }

        return jsonTypeInfo;
    }
}
```

## Additional details

* Polymorphic serialization only supports derived types that have been explicitly opted in via the `JsonDerivedType` attribute. Undeclared runtime types will result in a runtime exception. The behavior can be changed by configuring the `JsonPolymorphicAttrihute.UnknownDerivedTypeHandling` property.
* Polymorphic configuration specified in derived types is not inherited by polymorphic configuration in base types. These need to be configured independently.
* Polymorphic hierarchies are supported for both classes and interface types. 
* Polymorphism using type discriminators is only supported for type hierarchies that use the default converters for objects, collections and dictionary types.
* Polymorphism is supported in metadata-based sourcegen, but not fast-path sourcegen.

## API Proposal

```C#
namespace System.Text.Json.Serialization
{
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Interface, AllowMultiple = false, Inherited = false)]
    public sealed class JsonPolymorphicAttribute : JsonAttribute
    {
        public string? TypeDiscriminatorPropertyName { get; set; }
        public bool IgnoreUnrecognizedTypeDiscriminators { get; set; }
        public JsonUnknownDerivedTypeHandling UnknownDerivedTypeHandling { get; set; }
    }

    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Interface, AllowMultiple = true, Inherited = false)]
    public class JsonDerivedTypeAttribute : JsonAttribute
    {
        public JsonDerivedTypeAttribute(Type derivedType);
        public JsonDerivedTypeAttribute(Type derivedType, string typeDiscriminatorId);
        public JsonDerivedTypeAttribute(Type derivedType, int typeDiscriminatorId);

        public Type DerivedType { get; }
        public object? TypeDiscriminatorId { get; }
    }

    public enum JsonUnknownDerivedTypeHandling
    {
        FailSerialization = 0, // Default, fail serialization on undeclared derived type
        FallBackToBaseType = 1, // Fall back to the base type contract
        FallBackToNearestAncestor = 2 // Fall back to the nearest declared derived type contract (admits diamond ambiguities in cases of interface hierarchies)
    }
}
```
```C#
namespace System.Text.Json.Metadata;

public partial class JsonTypeInfo
{
    public JsonPolymorphismOptions? PolymorphismOptions { get; set; } = null;
}

public class JsonPolymorphismOptions
{
    public JsonPolymorphismOptions();

    public ICollection<JsonDerivedType> DerivedTypes { get; }

    public bool IgnoreUnrecognizedTypeDiscriminators { get; set; } = false;
    public JsonUnknownDerivedTypeHandling UnknownDerivedTypeHandling { get; set; } = JsonUnknownDerivedTypeHandling.Default;
    public string TypeDiscriminatorPropertyName { get; set; } = "$type";
}

// Use a dedicated struct instead of ValueTuple that handles type checking of the discriminator id
public struct JsonDerivedType
{
    public JsonDerivedType(Type derivedType);
    public JsonDerivedType(Type derivedType, int typeDiscriminatorId);
    public JsonDerivedType(Type derivedType, string typeDiscriminatorId);

    public Type DerivedType { get; }
    public object? TypeDiscriminatorId { get; }
}
```

## Anti-Goals

* No support for serializing open hierarchies (as specified in #29937).
* No support for deserializing open hierarchies.

## Progress

- [x] Bring [prototype branch](https://github.com/dotnet/runtime/pull/53882) up to date.
- [x] Implementation & testing.
- [x] API proposal & review.
- [ ] Conceptual documentation & blog posts.

**Comment by eiriktsarpalis** [comment]
> If it simply means "Should we allow users to serialize a collection of objects of various types that all derive from IType?" (e.g. List) , then I can't possibly imagine a scenario where the answer would be no.

That's right. The issue primarily is that collections serialize as JSON arrays. As such introducing a type discriminator would serve to complicate the contract, i.e. it would likely require nesting the array payload inside an object with metadata:
```json
{ "$type" : "derivedList", "$values" : [1,2,3,4,5] }
```

**Comment by SCLDGit** [comment]
It does feel a bit insulting to have a much-requested feature completely disallowed simply because people _might_ misuse it. Even if we had to toss it in an unsafe block, it would be nice to have simple does-what-it-says-on-the-box JSON serialization/deserialization for the (very common) case where security is not a concern. I've used JSON as a quick, simple data interchange format (most recently to store scene data for rendering software) more times than I can count, and not having to rely on Newtonsoft.JSON would be great for those cases.

**Comment by eiriktsarpalis** [comment]
> For the above examples, would it be too much to also support an inverse of the attribute:

Such an attribute would likely need to explicitly specify the base type, since otherwise that would be ambiguous: was it meant for `Base`? or `object`? or some interface that `Derived1` happens to implement? 

In any case, it should be possible to work around not owning the base type by specifying the hierarchy via `JsonSerializerOptions`.

**Comment by jeffhandley** [comment]
We acknowledge and understand the frustration that folks have with open hierarchies being disallowed. Our intent with publishing these goals and anti-goals is to help bring clarity to expectations and engage in these conversations. This should help us find a viable path forward together.

> It does feel a bit insulting to have a much-requested feature completely disallowed simply because people _might_ misuse it.

The challenge with open hierarchies isn't that people _might_ misuse it, it's that _any use_ of it is insecure. There are scenarios in which that vulnerability is masked by defenses _elsewhere_ in a system, including the "I wrote this code and it's only ever going to run on my machine with my own data" scenario, but we cannot design this feature set with reliance on outside defenses. We must approach this design knowing that foundational security is paramount and then enable as many scenarios as we can with features that are still usable. We want to be clear about how we're approaching the problem space.

By designing with security first, we can keep building new capabilities outward to incrementally shrink the set of scenarios that cannot be accomplished. But if we start with features that are vulnerable by design, then we'll inevitably reach a point where it's difficult to use the features _securely_. That was the fate of `BinaryFormatter`, which is planned to be [_removed_ from .NET](https://github.com/dotnet/designs/blob/main/accepted/2020/better-obsoletion/binaryformatter-obsoletion.md) as a result.

We want to hear more about scenarios where open hierarchies have been used, and explore approaches for addressing those scenarios without relying on open hierarchies as an implementation detail. Thank you for helping us keep this moving forward!

**Comment by eiriktsarpalis** [comment]
And, for what it's worth, it should still be possible to implement unconstrained polymorphism using custom converters. I've seen a few examples like that on the internet (including a couple of threads in this repo), although I wouldn't want to share links to them here.

**Comment by Symbai** [comment]
> And, for what it's worth, it should still be possible to implement unconstrained polymorphism using custom converters.

But we don't want hand crafted solutions. If we would, the request wouldn't have so many up votes where people are telling you (and everyone posting his custom solution) this over and over again. The problem is rather simple: If the JSON .NET team make any change in the future because they are not aware of everyone's custom solution, and this change breaks this custom solution. But this custom solution has been used for so long that so many JSON (files for example) already exist. Then we have the worst case scenario ever. As a developer we want reliability, we want something where we can be sure it also works in 5 years or more. And we want it simply, something to opt-in. Something that works everywhere the same. No matter which developer wrote the code for it.

Newtonsoft JSON provides it. System.Text.JSON does not. And since .NET 3.1 we are asking for it and now that it has been delayed for so many times telling us to use custom converters is a bit odd. If that was ever an option, we would have used it already. But we are not, we are waiting for an official solution. Please keep this in mind, thank you.

**Comment by eiriktsarpalis** [comment]
> But we are not, we are waiting for an official solution.

Just to be absolutely clear, there will not be an officially supported way to do unconstrained polymorphic serialization in the foreseeable future. We are always open to offering better extensibility points that make such features easier to build as extensions, but the onus for doing this will always be on the developer/ecosystem.

**Comment by SCLDGit** [comment]
This is an especially annoying situation for those of us on the desktop app side of the equation who use JSON as a data storage or round-trip save format. The only security concern to us, at least as far as I understand it, is that someone could alter an on-disk JSON file that our software has generated that could do something malicious when loaded back into the system. If this is indeed the case (and please correct me if it's not), then there are already myriad data integrity validation tools at our disposal, including cryptographic hashing and singing, that make this concern a non issue. In every scenario where I want to serialize and deserialize JSON, I am in control of the data from end to end. Ignoring this use case in favor of only addressing the concerns of anonymous web transfers is ignoring a large chunk of the equation. No one is suggesting that security not be the prime concern by default, but if every library design approached its domain problem with the idea that developers should be absolutely disallowed from getting ourselves into potential trouble, even if it's opt-in, nothing would ever get done in the field of software development. Please reconsider this stance. Many of us don't require this level of hand holding.

**Comment by jeremyVignelles** [comment]
I agree with the fact that unconstrained polymorphism is a security issue, and as far as I understand, the only real counter-arguments here are that it worked before and that it's easy to dump process memory and load it back.

If it can be done with the proper extensions method, and that it's as useful as you say, I'm sure there will be an open-source package that will fill the gap, without tempting new devs that don't understand the security implications of this. What do you think?

**Comment by ishepherd** [comment]
bq. I'm sure there will be an open-source package that will fill the gap,

This seems logical. As S.T.J necessarily evolves slowly, an open source extensions package it is a natural idea. S.T.J should provide good enough extension points to let this package be built.

Because the package is not by MS, MS cannot rightly be blamed for the vulns it introduces.

**Comment by eiriktsarpalis** [comment]
> provide the serialization facilities in ways that do not enforce class invariants

Can you clarify a bit more what you mean by class invariants?

**Comment by Symbai** [comment]
> If it can be done with the proper extensions method, and that it's as useful as you say, I'm sure there will be an open-source package that will fill the gap

There are already open source packages which provide polymorphism support. But I found they don't work in all cases or aren't updated anymore. This is what can(!) happen to all third party packages. If you rely on them and got thousands of JSON files and it doesn't work anymore and you have to use another package perhaps and all of your previous JSON becomes invalid.. good luck. Not to say that any further updates to System.Text.Json can break it as well. I don't think this an appropriated solution for a serious developer/company but that's just my opinion.

**Comment by quixoticaxis** [comment]
> > provide the serialization facilities in ways that do not enforce class invariants
> 
> Can you clarify a bit more what you mean by class invariants?

@eiriktsarpalis
I mean the set of assertions that stay true through the lifetime of the object, often enforced by the constructors.

I can understand that the class that provides public setters with no validation for properties of non specific types can be insecurely deserialized even without polymorphism, but I fail to imagine an example of a class that 1) would be insecure to recreate on deserialization through constructor, and 2) would not have other obvious design issues.
A quick Internet search didn't help.

**Comment by eiriktsarpalis** [comment]
I would say that's an orthogonal concern. Polymorphism is achievable while still honoring class invariants in the type hierarchy. TL;DR the security concern around polymorphic deserialization is allowing the payload to specify the arbitrary .NET type it wants to deserialize into, which can be exploited for remote code execution. A few real examples of such exploits can be found [here](https://github.com/pwntester/ysoserial.net).

**Comment by quixoticaxis** [comment]
> I would say that's an orthogonal concern. Polymorphism is achievable while still honoring class invariants in the type hierarchy. TL;DR the security concern around polymorphic deserialization is allowing the payload to specify the arbitrary .NET type it wants to deserialize into, which can be exploited for remote code execution. A few real examples of such exploits can be found [here](https://github.com/pwntester/ysoserial.net).

Oh, I see, we're speaking about the scenario when the code author is clearly stating the intent to handle literally anything by deserializing `object`, but does not think about the consequences. Is it really an issue?
One of the issues with binary deserialization is the ability to re-create a thoughtfully designed object of some type that does not omit validation in a state that is not reachable throughout in-memory object's lifetime. The deserialization through constructors is completely devoid of this flaw, or at least puts the burden of security (via validationon of inputs) on the class' author.
Maybe I'm missing something, but I'm not certain I agree with lots of people (who are trying to `Deserialize<MyBase>(payload)`) paying the price of additional code bloat to secure that one guy who decided to "deserialize ANY object, `cause what can go wrong?"

**Comment by eiriktsarpalis** [comment]
It's not just `object` though, it could be any base class or interface type that is widely used enough (a.k.a. the framework could contain subtypes that are unknown to the developer). What's more, this is not a concern restricted to the root type, for instance in your example `MyBase` could have a derived type defining a property of type `object`, which should be sufficient to write a gadget triggering remote code execution.

**Comment by quixoticaxis** [comment]
> It's not just `object` though, it could be any base class or interface type that is widely used enough (a.k.a. the framework could contain subtypes that are unknown to the developer). What's more, this is not a concern restricted to the root type, for instance in your example `MyBase` could have a derived type defining a property of type `object`, which should be sufficient to write a gadget triggering remote code execution.
> 
> My point is creating a vulnerable deserialization model is easier than what folks might initially anticipate. Individual developers might "know what they are doing" and take great care to avoid such pitfalls, however this is harder to enforce in the context of ever-changing teams working on a shared codebase. It's better if the tool enforces this by design, but folks should still be able to use a third-party extension if they like living dangerously.

Yeah, obviously any base that has "dangerous" derived classes readily available would work.

Point taken.

**Comment by mihael74** [comment]
I think I have found a workaround for now: Just put a JsonPropertyOrderAttribute to a property in every hierarchy an all  hierarchies are serialized.
Of course this is just a workaround.

**Comment by eiriktsarpalis** [comment]
I have updated the original post with what I believe is going to be the finalized shape and semantics of the feature. Please take a look, your feedback would be greatly appreciated.

**Comment by krwq** [comment]
I have some naming suggestions:
- `JsonDerivedType` => `JsonAllowDerivedType`
- `typeDiscriminatorId` => `typeIdentifier`

also couple of questions:
- how to specify type identifier for base class (assuming base class can be instantiated directly) or no identifier means base?
- how to specify allowed derived types for types outside of your control (assembly in someone else's nuget package - only through contract resolve?
- will the attribute work for second level of inheritance if attribute is put in the `between` class?

**Comment by SCLDGit** [comment]
Some kind of [JsonShouldDerive] tag for base classes would be nice for writing IDE tooling to automatically generate [JsonDerivedType] tags when creating inheriting classes, or to have Intellisense (or similar) recommend it as an action. Anything to make proper tagging less prone to human error would be a big plus.

**Comment by eiriktsarpalis** [comment]
> how to specify type identifier for base class (assuming base class can be instantiated directly) or no identifier means base?

It would still be possible to specify a discriminator for the base class (using the same `JsonDerivedType` mechanism for the base type itself).

> how to specify allowed derived types for types outside of your control (assembly in someone else's nuget package - only through contract resolve?

Correct -- only the contract resolver would be able to unblock this scenario.

> will the attribute work for second level of inheritance if attribute is put in the "between" class?

No. If multiple types in a hierarchy specify polymorphism attributes these specify configuration for the given declared type only.

**Comment by krwq** [comment]
@SCLDGit once you fully automate it you're super prone to remote code execution on de-serialization :-) It's purposefully made harder so you use it only when you have to

**Comment by SCLDGit** [comment]
@krwq I'm not talking full automation, I'm more talking something where if I define a `Base` class and then `Derived : Base`, and I have base decorated with something like `[JsonShouldDerive]`, the IDE can recommend an action to generate the `[JsonDerivedType(typeof(Derived), "derived")]` tag on the parent class automatically.

**Comment by steveharter** [comment]
As an aside, it may be useful to enumerate the current polymorphic-related toolset in STJ:
- This new polymorphic feature that has a closed hierarchy with embedded, safe metadata (i.e. the metadata does not contain the raw Type names to instantiate).
- Serializer:
  - The custom converter feature; this can become cumbersome especially if the custom logic is to respect the various knobs on `JsonSerializerOptions`.
  - Support for declaring a property as `object` which supports polymorphic serialization of that property (i.e. `object.GetType()` is used).
    - Similarly, there are `JsonSerializer.Serialize()` APIs that take a `Type` parameter instead of a generic parameter. This allows a call to `object.GetType()` to be made prior to `Serialize` in order to polymorphically serialize that root object or limit to a base class.
- DOM support with `JsonElement` and `JsonNode`:
  - A property declared as `object` is deserialized as a `JsonElement` by default but can be configured to use `JsonNode` via `options.JsonUnknownTypeHandling = JsonUnknownTypeHandling.JsonNode`.
    - The 6.0 release added `JsonNode` that supports edits; `JsonElement` is read-only. This opens up new scenarios since it allows for easier round-tripping of modifiable properties.
    - Note that properties can be declared with `JsonElement` or `JsonNode` directly instead of `object`, including on generic collections.
  - The "overflow" feature using `[JsonExtensionData]`. This allows for unknown properties (that don't exist on the target object) to be stored in a dictionary-backed property during deserialization as `JsonElement` or `JsonNode` which are then serialized later. This is useful for versioning as well as loosely-typed polymorphic scenarios.
  - Support for easily converting to\from  DOM types (`JsonElement` and `JsonNode`) and objects. i.e. `JsonSerializer.Deserialize<T>(JsonNode)` and `JsonSerializer.SerializeToNode()`.

**Comment by eiriktsarpalis** [comment]
Good point, there's also a fair bit of documentation that will be going out of date once this is ready, e.g. https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-polymorphism

**Comment by eiriktsarpalis** [comment]
With #63686 approved I have now updated the API proposal to its final form and marked the issue ready for review. Please take a look -- any feedback ahead of API review would be greatly appreciated.

**Comment by Timovzl** [comment]
@eiriktsarpalis Nitpick: For consistency, since `JsonUnknownDerivedTypeHandling.FailSerialization` is based on a verb ("fail"), `FallbackTo*` probably be `FallBackTo*` (capital B). That way, it's based on the verb "fall back" rather than the noun "[a] fallback".

**Comment by eiriktsarpalis** [comment]
@Timovzl good feedback, I've updated the proposal.

**Comment by terrajobst** [comment]
* `JsonPolymorphicAttribute`
    - `CustomTypeDiscriminatorPropertyName` -> `TypeDiscriminatorPropertyName`
* `JsonPolymorphicTypeInfo` sounds like it should be derived from `JsonTypeInfo`
    - `CustomTypeDiscriminatorPropertyName` -> `TypeDiscriminatorPropertyName`
    - The class itself should probably not implement `ICollection`
    - It should, however, have a property or method to enumerate the mapping
    - We'd prefer a design without the fluent `WithXxx()` methods and instead use a void returning `Add` method.
* Looks like we don't need `JsonPolymorphismOptions<TBaseType>`


```C#
namespace System.Text.Json.Serialization;

[AttributeUsage(AttributeTargets.Class |
                AttributeTargets.Interface,
                AllowMultiple = false,
                Inherited = false)]
public sealed class JsonPolymorphicAttribute : JsonAttribute
{
    public string? CustomTypeDiscriminatorPropertyName { get; set; }
    public bool IgnoreUnrecognizedTypeDiscriminators { get; set; }
    public JsonUnknownDerivedTypeHandling UnknownDerivedTypeHandling { get; set; }
}

[AttributeUsage(AttributeTargets.Class |
                AttributeTargets.Interface,
                AllowMultiple = true,
                Inherited = false)]
public class JsonDerivedTypeAttribute : JsonAttribute
{
    public JsonDerivedTypeAttribute(Type derivedType);
    public JsonDerivedTypeAttribute(Type derivedType, string typeDiscriminatorId);
    public JsonDerivedTypeAttribute(Type derivedType, int typeDiscriminatorId);

    public Type DerivedType { get; }
    public object? TypeDiscriminatorId { get; }
}

public enum JsonUnknownDerivedTypeHandling
{
    FailSerialization = 0,
    FallBackToBaseType = 1,
    FallBackToNearestAncestor = 2
}
```

```C#
namespace System.Text.Json.Metadata;

public partial class JsonTypeInfo
{
    public JsonPolymorphismOptions? PolymorphicTypeInfo { get; set; } = null;
}

public class JsonPolymorphismOptions : ICollection<(Type DerivedType, object? TypeDiscriminatorId)>
{
    public JsonPolymorphismOptions(Type baseType);
    public Type BaseType { get; }
    public string? CustomTypeDiscriminatorPropertyName { get; set; }
    public bool IgnoreUnrecognizedTypeDiscriminators { get; set; }
    public JsonUnknownDerivedTypeHandling UnknownDerivedTypeHandling { get; set; }
    public JsonPolymorphismOptions WithDerivedType(Type derivedType);
    public JsonPolymorphismOptions WithDerivedType(Type derivedType, int typeDiscriminatorId);
    public JsonPolymorphismOptions WithDerivedType(Type derivedType, string typeDiscriminatorId);
}
```

>

**Comment by eiriktsarpalis** [comment]
Given review feedback, I'd be inclined to flatten polymorphism configuration into properties of `JsonTypeInfo` proper. This would look as follows:
```C#
namespace System.Text.Json.Metadata;

public partial class JsonTypeInfo
{
    public string TypeDiscriminatorPropertyName { get; set; } = "$type";
    public bool IgnoreUnrecognizedTypeDiscriminators { get; set; } = false;
    public JsonUnknownDerivedTypeHandling UnknownDerivedTypeHandling { get; set; } = JsonUnknownDerivedTypeHandling.Default;

    [AllowNull]
    public JsonDerivedTypes DerivedTypes { get; set; }
}

public class JsonDerivedTypes : ICollection<JsonDerivedType>
{
     public JsonDerivedTypes();

     // facilitate collection initializer syntax
     public void Add(JsonDerivedType derivedType); 

     // optional accelerator methods, note that these can introduce ambiguity in collection initializers,
     // e.g. we can no longer target-type `JsonDerivedType` constructors or use overloads with two parameters.
     // We might want to consider renaming the accelerator methods to something different or scrap them altogether.
     public void Add(Type derivedType);
     public void Add(Type derivedType, int typeDiscriminatorId);
     public void Add(Type derivedType, string typeDiscriminatorId);
}

// Use a dedicated struct instead of ValueTuple that handles type checking of the discriminator id
public struct JsonDerivedType
{
    public JsonDerivedType(Type derivedType);
    public JsonDerivedType(Type derivedType, int typeDiscriminatorId);
    public JsonDerivedType(Type derivedType, string typeDiscriminatorId);

    public Type DerivedType { get; }
    public object? TypeDiscriminatorId { get; }
}
```
Under this design, a type metadata would be regarded as polymorphic iff its `JsonDerivedTypes` collection is nonempty. 

## Usage Examples

```C#
// 1. Specifying configuration using 
jsonTypeInfo.TypeDiscriminatorPropertyName = "_case";
jsonTypeInfo.UnknownDerivedTypeHandling = JsonUnknownDerivedTypeHandling.FallBackToNearestAncestor;
jsonTypeInfo.DerivedTypes.Add(typeof(DerivedType1));
jsonTypeInfo.DerivedTypes.Add(typeof(DerivedType2), "derivedType2");
jsonTypeInfo.DerivedTypes.Add(typeof(DerivedType3), 42);

// 2. Using collection initializers
jsonTypeInfo.DerivedTypes = new JsonDerivedType()
{
    new JsonDerivedType(typeof(DerivedType1)),
    new JsonDerivedType(typeof(DerivedType2), "derivedType2"),
    new JsonDerivedType(typeof(DerivedType3), 42),
};

// 3. Clear any polymorphism metadata inherited from attributes
jsonTypeInfo.DerivedTypes = null; 
```
cc @krwq @layomia @bartonjs for additional feedback.

**Comment by bartonjs** [comment]
Flattening the options down onto JsonTypeInfo feels a little weird to me... I don't see any other "clustered" settings on it.  That is, `TypeDiscriminatorPropertyName` and `UnknownDerivedTypeHandling` have no meaning unless `DerivedTypes` is non-empty (or non-null?)... so now they're either error states or just garbage data.

If we think that 90%+ of the user usage is going to be attributes (either reflection-based serialization or source generation) then it doesn't matter a whole bunch... but, still, it feels a bit off to me.

**Comment by krwq** [comment]
I got mixed feelings on flattening: examples above from @eiriktsarpalis seem much cleaner than no flatenning and at the same they are grouped. But having said that only UnknownDerivedTypeHandling feels a bit off but at the same time NumberHandling is also meaningless unless Type is numeric

**Comment by eiriktsarpalis** [comment]
## Updated proposal

```C#
namespace System.Text.Json.Metadata;

public partial class JsonTypeInfo
{
    public JsonPolymorphismOptions? PolymorphismOptions { get; set; } = null;
}

public class JsonPolymorphismOptions
{
    public JsonPolymorphismOptions();

    public ICollection<JsonDerivedType> DerivedTypes { get; }

    public bool IgnoreUnrecognizedTypeDiscriminators { get; set; } = false;
    public JsonUnknownDerivedTypeHandling UnknownDerivedTypeHandling { get; set; } = JsonUnknownDerivedTypeHandling.Default;
    public string TypeDiscriminatorPropertyName { get; set; } = "$type";
}

// Use a dedicated struct instead of ValueTuple that handles type checking of the discriminator id
public struct JsonDerivedType
{
    public JsonDerivedType(Type derivedType);
    public JsonDerivedType(Type derivedType, int typeDiscriminatorId);
    public JsonDerivedType(Type derivedType, string typeDiscriminatorId);

    public Type DerivedType { get; }
    public object? TypeDiscriminatorId { get; }
}
```

### Usage Examples

```C#
// 1. Specifying new polymorphic configuration
jsonTypeInfo.PolymorphismOptions =
    new JsonPolymorphismOptions
    {
         TypeDiscriminatorPropertyName = "_case",
         UnknownDerivedTypeHandling = JsonUnknownDerivedTypeHandling.FallBackToNearestAncestor,
         DerivedTypes =
         {
              new JsonDerivedType(typeof(DerivedType1)),
              new JsonDerivedType(typeof(DerivedType2), "derivedType2"),
              new JsonDerivedType(typeof(DerivedType3), 42),
         }
    };

// 2. Clear any polymorphism metadata inherited from attributes
jsonTypeInfo.PolymorphismOptions.TypeDiscriminatorId = "_case";
jsonTypeInfo.DerivedTypes.Clear();
jsonTypeInfo.DerivedTypes.Add(new JsonDerivedType(typeof(DerivedType1));
```

**Comment by terrajobst** [comment]
* Let's drop the `Id` suffix of `TypeDiscriminator`

```C#
namespace System.Text.Json.Serialization;

[AttributeUsage(AttributeTargets.Class |
                AttributeTargets.Interface,
                AllowMultiple = false,
                Inherited = false)]
public sealed class JsonPolymorphicAttribute : JsonAttribute
{
    public string? TypeDiscriminatorPropertyName { get; set; }
    public bool IgnoreUnrecognizedTypeDiscriminators { get; set; }
    public JsonUnknownDerivedTypeHandling UnknownDerivedTypeHandling { get; set; }
}

[AttributeUsage(AttributeTargets.Class |
                AttributeTargets.Interface,
                AllowMultiple = true,
                Inherited = false)]
public class JsonDerivedTypeAttribute : JsonAttribute
{
    public JsonDerivedTypeAttribute(Type derivedType);
    public JsonDerivedTypeAttribute(Type derivedType, string typeDiscriminator);
    public JsonDerivedTypeAttribute(Type derivedType, int typeDiscriminator);

    public Type DerivedType { get; }
    public object? TypeDiscriminator { get; }
}

public enum JsonUnknownDerivedTypeHandling
{
    FailSerialization = 0,
    FallBackToBaseType = 1,
    FallBackToNearestAncestor = 2
}
```

```C#
namespace System.Text.Json.Metadata;

public partial class JsonTypeInfo
{
    public JsonPolymorphismOptions? PolymorphismOptions { get; set; }
}

public class JsonPolymorphismOptions
{
    public JsonPolymorphismOptions();

    public ICollection<JsonDerivedType> DerivedTypes { get; }

    public bool IgnoreUnrecognizedTypeDiscriminators { get; set; }
    public JsonUnknownDerivedTypeHandling UnknownDerivedTypeHandling { get; set; }
    public string TypeDiscriminatorPropertyName { get; set; }
}

public struct JsonDerivedType
{
    public JsonDerivedType(Type derivedType);
    public JsonDerivedType(Type derivedType, int typeDiscriminator);
    public JsonDerivedType(Type derivedType, string typeDiscriminator);

    public Type DerivedType { get; }
    public object? TypeDiscriminator { get; }
}
```

**Comment by BreyerW** [comment]
A question how configurating polymorphism for non-owned types work in case of metadata source gen? AFAIK it was suggested that custom contract resolver will NOT be picked up by source gen path rendering `JsonPolymorphimsOptions` nonviable for sourcegen. Or it is nonissue by virtue of static nature of source gen (that is, it knows all types that are to be serialized and thus can generate exhaustive discriminator mappings on its own)?

**Comment by eiriktsarpalis** [comment]
> A question how configurating polymorphism for non-owned types work in case of metadata source gen? AFAIK it was suggested that custom contract resolver will NOT be picked up by source gen path rendering `JsonPolymorphimsOptions` nonviable for sourcegen. Or it is nonissue by virtue of static nature of source gen (that is, it knows all types that are to be serialized and thus can generate exhaustive discriminator mappings on its own)?

Good question. There currently is no way make the source generator opt-in non-owned types into polymorphism, but it should still be possible to define a custom contract resolver deriving from JsonSerializerContext:

```C#
[JsonSerializable(typeof(NonOwnedBaseType))]
[JsonSerializable(typeof(NonOwnedDerivedType))]
public MyContext : JsonSerializerContext
{
}

public class MyContractResolver : IJsonTypeInfoResolver
{
     public JsonTypeInfo GetTypeInfo(Type type, JsonSerializerOptions options)
     {
          JsonTypeInfo sourceGenTypeInfo = ((IJsonTypeInfoResolver)MyContext.Default).GetTypeInfo(type, options);

          if (type == typeof(NonOwnedBaseType))
          {
               sourceGenTypeInfo.PolymorphismOptions = /* add polymorphism metadata as usual */
          }

          return sourceGenTypeInfo;
     }
}

var options = new JsonSerializerOptions { TypeInfoResolver = new MyContractResolver() };
JsonSerializer.Serialize(new NonOwnedDerivedType(), options);
```
In the future we might consider exposing polymorphism configuration options via the `JsonSourceGenerationOptionsAttribute` to make the experience better for source gen scenaria.

**Comment by BreyerW** [comment]
@eiriktsarpalis  hmm if sourcegen can actually pick up contracts and afaik public contract design with polymorphism is still in flight maybe there should be public static (or singleton) `PolymorphismOptions` mapping to which sourcegen would populate mapping based on attributes (also user could add programmatically mappings of lower priority than contract's mapping) and in actual serialization it would look up contract's mapping first (which would be empty in most cases) then public static mapping (which would be mostly filled with attribute based mapping).

 Filling of static/singleton options could be done in generated module initializer or require custom contracts to be partial for source gen and generate population there (maybe in constructor or `Initialize` method). The latter option would let us skip static/singleton options and in case someone wanted default contract with polymorphism then they can inhert default contract resolver with empty, partial class and register it, introducing minimal boilerplate for this scenario

The idea is to ensure theres no two separate ways for reflection and source gen serialization to enable non-owned types in polymorphism, which would ideally be achieved by fully sharing contract's interface between reflection and sourcegen mode

**Comment by Hawxy** [comment]
One limitation a noticed with the current implementation, it seems that `$type` is ordering sensitive. 

If you're using `jsonb` PostgresSQL columns such as with EFCore or MartenDB, keys can end up in any order the database considers ideal for performance.

For example:

```cs
var databaseState = @"{
  ""BaseDictionary"": {
    ""6ed6e524-2ca4-4fea-8e21-7245ccb61863"": {
      ""Id"": ""6ed6e524-2ca4-4fea-8e21-7245ccb61863"",
      ""Name"": ""Something"",
      ""$type"": ""Derived"",
      ""OtherGuid"": ""d471c77d-5412-4e7a-a98d-8304e87792ed""
    }
  }
}";

JsonSerializer.Deserialize<WrapperType>(databaseState);

public record WrapperType(Dictionary<Guid, WrapperType.Base> BaseDictionary)
{
    [JsonDerivedType(typeof(Derived), nameof(Derived))]
    [JsonDerivedType(typeof(AlsoDerived), nameof(AlsoDerived))]
    public abstract record Base(Guid Id, string Name);
    public record Derived(Guid Id, string Name, Guid OtherGuid): Base(Id, Name);
    public record AlsoDerived(Guid Id, string Name) : Base(Id, Name);

}
```
Results in:
```
System.NotSupportedException: 'Deserialization of types without a parameterless constructor, a singular parameterized constructor, or a parameterized constructor annotated with 'JsonConstructorAttribute' is not supported. Type 'WrapperType+Base'. Path: $.BaseDictionary.6ed6e524-2ca4-4fea-8e21-7245ccb61863 | LineNumber: 3 | BytePositionInLine: 11.'
```

Newtonsoft had a `MetadataPropertyHandling.ReadAhead` setting to get around this, but I don't see anything similar for STJ.

**Comment by ishepherd** [comment]
@Hawxy ugh, is that the behaviour? That seems broken.

**Comment by eiriktsarpalis** [comment]
@Hawxy yes, this is a known limitation of the JSON metadata reader implementation (also used in features like reference preservation). Without getting into too much detail, it boils down to the fact that the internal converters have been written with async streaming in mind, as such the type discriminator might not even have been loaded in memory when the first properties are being read. In the future we might consider exposing an optional flag that works around this, at the expense of   reading the same JSON twice and buffering excessive data in memory.

**Comment by BreyerW** [comment]
@eiriktsarpalis Could you answer my question i made in edit? Need to understand actual limitations since im thinking of moving away from Newtonsoft to STJ and last obstacles are polymorohism and publicly exposed contract (well and support for private members ar least just fields but i think i will be able to workaround it)

> EDIT: thinking about it further i might misunderstood you (english isnt my main language sorry). Did you mean only scenario where we want to enable non-owned types without custom contract isnt supported? But we can still add them manually via custom contract and avoiding custom contract would be enabled by future JsonSourceGenerationOptionsAttribute? At first I was under impression that sourceGenTypeInfo.PolymorphismOptions would be noop in sourcegen scenario but im not sure anymore ... Cant play with preview system.text.json and sourcegen for now so cant verify

**Comment by eiriktsarpalis** [comment]
> At first I was under impression that sourceGenTypeInfo.PolymorphismOptions would be noop in sourcegen scenario but im not sure anymore ... Cant play with preview system.text.json and sourcegen for now so cant verify

The only difference between source gen and reflection is that the former generates metadata at compile-time while the latter uses reflection. The core serialization logic is agnostic as to the source of metadata so if `PolymorphismOptions` is specified it would be honored.

**Comment by BreyerW** [comment]
@eiriktsarpalis in that case i was indeed under wrong impression and everything is okay with `JsonSourceGenerationOptionsAttribute` idea (but also feels kinda redundant if default resolver wont be sealed and be public and can be extended). Thanks for clarification.

**Comment by onionhammer** [comment]
One thing that my library appears to have over this is compile time build errors for enum valued type discriminators.. not sure how this proposal handles enums here, the samples rely on ints and magic strings

**Comment by AignerGames** [comment]
I like the idea, but after reading this I have a few questions, because Im not sure if / what will be implemented of the following things:

- Will this system also work for generic types (or interfaces), for example IMyGenericInterface<T> which is implemented by MyGenericClass<T> or MyStringClass<string> and uses as a class property of type public ISomeGeneric<string> MyStringProperty {get; set;}

- Adding all the attribute for every "valid" type by hand seems a bit annoying, everytime you add a new class you also have to remember to add this attribute for the new derived type, is there no way to "automatically" detect and fill the list? I mean I can use reflection myself to detect all "valid" types, but how can I tell your system about this list? I can't add the attribute at runtime (correct me if Im wrong there)

The JsonTypeInfo examples posted above seem like a good option for that case, because then I can collect all "valid" types myself, even at runtime, and just add the info, but Im not sure if this is / will actually be added or was that just an example?


Ok now for the dumb question; I still don't fully understand the security issue, if I add a type property to the json like "Type" : "MyDerivedType" and the property is of type "MyBaseType", how is this a issue? It will find MyDerivedType and deserialize it, but only if it is an actual derived type of "MyBaseType" (it's easy to check). So it shouldn't be a issue, because MyDerivedType will only exist and be derived from MyBaseType if I added it to my code, it would only be an issue if I allow additional dlls to be loaded at runtime (but this has many other security issues anyways if not done with care) and someone adds a "bad MyEvilDerivedType" and then adds the type name in the json file, but unless I allow loading of additional dlls at runtime I don't see the issue and we we need all these "safety attributes"?

**Comment by eiriktsarpalis** [comment]
> Will this system also work for generic types (or interfaces)

Attribute annotations will not work for generic derived types, such configuration will need to happen using the contract resolver model. See the "Configuring Polymorphism via the Contract model" section in the OP.

> Adding all the attribute for every "valid" type by hand seems a bit annoying

Likewise, authoring a custom contract resolver should let you define polymorphic hierarchies programmatically.

> Ok now for the dumb question; I still don't fully understand the security issue, if I add a type property to the json like "Type" : "MyDerivedType" and the property is of type "MyBaseType", how is this a issue?

It becomes an issue when "MyBaseType" is high enough in the type hierarchy (e.g. `object` or `IEnumerable<T>`) to make it possible for attackers to force remote code execution using types already present in the BCL. It's not an issue of loading other assemblies as much as it is incorporating arbitrary types in the deserialization graph.

> Wouldn't it be a much simpler solution to just add a property to the (de)serializer settings with a list of "safe derived type" assembilies

That's precisely what contract customization would be making, only configuration would be scoped to polymorphic types rather than the assembly level.

**Comment by AignerGames** [comment]
Thank you for the answers @eiriktsarpalis, I must have skipped the contract model section before. I checked the example and it seems to be exactly what I need. 

I currently have a custom json serializer which collects all "valid" derived types from a specific "safe" assembly, which contains all the checks to find interface implementations and generic types etc. So it should be possible to simply change my implementaton to this the new feature and feed the custom contract model with my already existing code to collect the types.

**Comment by Michael-Bohin** [comment]
Thank you for the awesome work! I just copied one of the code snippets: 

`[JsonPolymorhic(UnknownDerivedTypeHandling = JsonUnknownDerivedTypeHandling.FallBackToNearestAncestor)]
[JsonDerivedType(typeof(MyDerivedClass)]
public interface IMyInterface { }
public class MyDerivedClass : IMyInterface { }

public class TestClass : MyDerivedClass { }

JsonSerializer.Serialize<IMyInterface>(new TestClass()); // serializes using the contract for `MyDerivedClass``

At the first line there is a missing 'p', you might want to fix it. The typo is actually in more of those snippets. Thanks again for this update, will use it heavily.

**Comment by ilya-scale** [comment]
The new model that you guys have introduced seems like a big improvement and is a very welcome addition! One thing I was wondering and cannot find an answer to is how a type discriminator would behave if we will have the exact same field that we want to use in a runtime:

Example:
```csharp
[JsonPolymorphic(CustomTypeDiscriminatorPropertyName = "Type")]
[JsonDerivedType(typeof(Derived))]
public class Base
{
    public string Type {get; }
    public int X { get; set; }
}

public class Derived : Base
{
    public int Y { get; set; }
}
```

What will happen if the `DiscriminatorPropertyName` will collide with an existing field and if it is possible to make it work "out of the box" that a field would both work as discriminator and a property.

Maybe this is really an anti-pattern to do it this way, and it is best to just check the C# type, but I have seen many places where code is using some additional Type field for various checks.

**Comment by eiriktsarpalis** [comment]
@ilya-scale that's a good question actually. It would seem like the metadata writer does not check for property name conflicts even though the metadata reader does and will throw an exception. I think it might make sense to introduce a fix so that some form of validation takes place on the serialization side as well. I'll fork this into a separate issue.

**Comment by bachratyg** [comment]
@ilya-scale this is a known limitation. See https://github.com/dotnet/runtime/issues/63747#issuecomment-1158624112

**Comment by Hawxy** [comment]
@eiriktsarpalis Is there an issue that tracks the above $type ordering limitation?

**Comment by eiriktsarpalis** [comment]
There isn't one. Would you like to create it?
